patches/server/0029-Actually-optimise-explosions.patch

```diff
@@ -0,0 +1,469 @@
++                return false;
++            }
++
++            // inc the smallest normalized coordinate
++
++            if (normalizedCurrX < normalizedCurrY) {
++                if (normalizedCurrX < normalizedCurrZ) {
++                    currX += dx;
++                    normalizedCurrX += normalizedDiffX;
++                } else {
++                    // x < y && x >= z <--> z < y && z <= x
++                    currZ += dz;
++                    normalizedCurrZ += normalizedDiffZ;
++                }
++            } else if (normalizedCurrY < normalizedCurrZ) {
++                // y <= x && y < z
++                currY += dy;
++                normalizedCurrY += normalizedDiffY;
++            } else {
++                // y <= x && z <= y <--> z <= y && z <= x
++                currZ += dz;
++                normalizedCurrZ += normalizedDiffZ;
++            }
++        }
++    }
++
++    private double getSeenFraction(final Vec3 source, final Entity target,
++                                   final ExplosionBlockCache[] blockCache,
++                                   final BlockPos.MutableBlockPos blockPos) {
++        final AABB boundingBox = target.getBoundingBox();
++        final double diffX = boundingBox.maxX - boundingBox.minX;
++        final double diffY = boundingBox.maxY - boundingBox.minY;
++        final double diffZ = boundingBox.maxZ - boundingBox.minZ;
++
++        final double incX = 1.0 / (diffX * 2.0 + 1.0);
++        final double incY = 1.0 / (diffY * 2.0 + 1.0);
++        final double incZ = 1.0 / (diffZ * 2.0 + 1.0);
++
++        if (incX < 0.0 || incY < 0.0 || incZ < 0.0) {
++            return 0.0;
++        }
++
++        final double offX = (1.0 - Math.floor(1.0 / incX) * incX) * 0.5 + boundingBox.minX;
++        final double offY = boundingBox.minY;
++        final double offZ = (1.0 - Math.floor(1.0 / incZ) * incZ) * 0.5 + boundingBox.minZ;
++
++        final io.papermc.paper.util.CollisionUtil.LazyEntityCollisionContext context = new io.papermc.paper.util.CollisionUtil.LazyEntityCollisionContext(target);
++
++        int totalRays = 0;
++        int missedRays = 0;
++
++        for (double dx = 0.0; dx <= 1.0; dx += incX) {
++            final double fromX = Math.fma(dx, diffX, offX);
++            for (double dy = 0.0; dy <= 1.0; dy += incY) {
++                final double fromY = Math.fma(dy, diffY, offY);
++                for (double dz = 0.0; dz <= 1.0; dz += incZ) {
++                    ++totalRays;
++
++                    final Vec3 from = new Vec3(
++                            fromX,
++                            fromY,
++                            Math.fma(dz, diffZ, offZ)
++                    );
++
++                    if (!this.clipsAnything(from, source, context, blockCache, blockPos)) {
++                        ++missedRays;
++                    }
++                }
++            }
++        }
++
++        return (double)missedRays / (double)totalRays;
```

Comment:
```suggestion
+        return (float)missedRays / (float)totalRays;
```
This is different compared to vanilla `getSeenPercent` and could cause some issues with TNT cannons.

Reply:
changed to adhere back to vanilla